src/snark/constraints.rs

```diff
@@ -0,0 +1,534 @@
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        let obj = f()?;
+
+        // Step 1: obtain the bits of the F field elements (little-endian)
+        let mut src_bits = Vec::<bool>::new();
+        for elem in obj.borrow().iter() {
+            let mut bits = elem.into_repr().to_bits();
+            // to_bits()'s result is big-endian
+
+            bits.reverse();
+            bits.truncate(F::size_in_bits());
+            for _ in bits.len()..F::size_in_bits() {
+                bits.push(false);
+            }
+            bits.reverse();
+
+            src_bits.append(&mut bits);
+        }
+
+        // Step 2: repack the bits as CF field elements
+        // Deciding how many bits can be embedded,
+        //  if CF has the same number of bits as F, but is larger,
+        //  then it is okay to put the entire field element in.
+        let capacity = if CF::size_in_bits() == F::size_in_bits() {
+            let fq = <<CF as PrimeField>::Params as FpParameters>::MODULUS;
+            let fr = <<F as PrimeField>::Params as FpParameters>::MODULUS;
+
+            let fq_u64: &[u64] = fq.as_ref();
+            let fr_u64: &[u64] = fr.as_ref();
+
+            let mut fq_not_smaller_than_fr = true;
+            for (left, right) in fq_u64.iter().zip(fr_u64.iter()).rev() {
+                if left < right {
+                    fq_not_smaller_than_fr = false;
+                    break;
+                }
+
+                if left > right {
+                    break;
+                }
+            }
+
+            if fq_not_smaller_than_fr {
+                CF::size_in_bits()
+            } else {
+                CF::size_in_bits() - 1
+            }
+        } else {
+            CF::size_in_bits() - 1
+        };
+
+        // Step 3: allocate the CF field elements as input
+        let mut src_booleans = Vec::<Boolean<CF>>::new();
+        for chunk in src_bits.chunks(capacity) {
+            let elem = CF::from_repr(<CF as PrimeField>::BigInt::from_bits(chunk)).unwrap(); // big endian
+
+            let elem_gadget = FpVar::<CF>::new_input(ns!(cs, "input"), || Ok(elem))?;
+
+            let mut booleans = elem_gadget.to_bits_le()?;
+            booleans.truncate(chunk.len());
+            booleans.reverse();
+
+            src_booleans.append(&mut booleans);
+        }
+
+        // Step 4: unpack them back to bits
+        let res = src_booleans
+            .chunks(F::size_in_bits())
+            .map(|f| {
+                let mut res = f.to_vec();
+                res.reverse();
+                res
+            })
+            .collect::<Vec<Vec<Boolean<CF>>>>();
+
+        Ok(Self {
+            val: res,
+            _snark_field_: PhantomData,
+        })
+    }
+}
+
+impl<F: PrimeField, CF: PrimeField> FromFieldElementsGadget<F, CF> for BooleanInputVar<F, CF> {
+    fn repack_input(src: &Vec<F>) -> Vec<CF> {
+        // the same routine as the new_input above
+
```

Comment:
If it's the same routine, then can we reuse the implementation to avoid duplication and things going out of sync?